### PR TITLE
Update double mapping decision to share memory

### PR DESCRIPTION
**Pull request by bragaigor**
Add helper to remapped leaves from shared to private
Make size of contiguous region multiple of region size

Depends on: https://github.com/eclipse/omr/pull/4920

Signed-off-by: Igor Braga <higorb1@gmail.com>

### CODE REVIEW

File: runtime/gc_glue_java/ArrayletObjectModel.cpp [hunk 1]
```diff
@@ -110,3 +115,38 @@ GC_ArrayletObjectModel::getArrayletLayout(J9Class* clazz, UDATA dataSizeInBytes,
 	}
 	return layout;
 }
+
+#if defined(J9VM_GC_ENABLE_DOUBLE_MAP)
+bool
+GC_ArrayletObjectModel::freeDoubleMapping(MM_EnvironmentBase *env, J9IndexableObject *objectPtr, struct J9PortVmemIdentifier *identifier)
```
Comment (amicic) [inline runtime/gc_glue_java/ArrayletObjectModel.cpp]: perhpas this makes more sense to reside in MM_IndexableObjectAllocationModel (where doubleMapArraylets() already is)? again, perhaps, think about it.

Comment (amicic) [reply runtime/gc_glue_java/ArrayletObjectModel.cpp]: or the other one should be here. bottom line, together in same class

File: runtime/gc_glue_java/ArrayletObjectModel.cpp [hunk 2]
```diff
@@ -110,3 +115,38 @@ GC_ArrayletObjectModel::getArrayletLayout(J9Class* clazz, UDATA dataSizeInBytes,
...
+			}
+			count++;
+			arrayletLeaveAddrs[count] = currentLeaf;
+		}
+		arrayletLeaveAddrs[0] = (void *)count;
+		identifier->handle = arrayletLeaveAddrs;
+	} else {
+		addressToFree = handle;
+	}
+	j9vmem_free_memory(identifier->address, identifier->size, identifier);
```
Comment (amicic) [inline runtime/gc_glue_java/ArrayletObjectModel.cpp]: hm.... I expected a call to updateDoubleMappedLeaves() sometime before this free()!?

Comment (amicic) [reply runtime/gc_glue_java/ArrayletObjectModel.cpp]: nor I see j9vmem_free_memory calling it

Comment (bragaigor) [reply runtime/gc_glue_java/ArrayletObjectModel.cpp]: `j9vmem_free_memory` calls `updateDoubleMappedLeaves()` in the OMR side as `omrvmem_free_memory` right at the end of the method.

File: runtime/gc_glue_java/ArrayletObjectModel.cpp (hunk 1 above)
Comment (bragaigor) [reply runtime/gc_glue_java/ArrayletObjectModel.cpp]: There was an issue when I tried to put `freeDoubleMapping` in `MM_IndexableObjectAllocationModel` as I could not find a clean way to call `freeDoubleMapping` from CopyForwardScheme.cpp, GlobalMarkingScheme.cpp, and PartialMarkingScheme.cpp. Maybe move `doubleMapArraylets` to ArrayletObjectModel.cpp?

Comment (amicic) [reply runtime/gc_glue_java/ArrayletObjectModel.cpp]: Ok, I see MM_IndexableObjectAllocationModel  is instantiated locally on stack during allocation, and does not exist any more at the end of GC when you need to unmap it.

It believe this one still could be a static method in MM_IndexableObjectAllocationModel (you don't need any of original constructor perameters  like class etc - it's all embedded into the object itself).
doubleMapArraylets  semantically belongs more  to AllocationModel IMO and it's good to leave it there.

Not sure after all - even as-is it's not too bad

Comment (bragaigor) [reply runtime/gc_glue_java/ArrayletObjectModel.cpp]: True, that would work, I'll check and see how that looks.

**Comment by amicic** [comment]
Jenkins test sanity xLinux jdk11

**Comment by amicic** [comment]
This change was required as part of effort to have file descriptor per each double mapped area. Meanwhile we decided to not go with this approach. Closing...

**Pull request closed without merging**